---
title: "[project_name] Peak QC"
author: "John Doe"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cosmo
    number_sections: true
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: show
    highlight: tango
    fig_caption: true
---


<script>
$(document).ready(function(){
    $('[data-toggle="popover"]').popover();
});
</script>

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE, 
                      echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE,
                      fig.align = 'center',
                      knitr.svg.object = TRUE,
                      dev=c('png', 'svglite', 'pdf'))
```

Load libraries:
```{r load-libraries}
# load library
library(tidyverse)
library(plyranges)
library(GenomicRanges)
library(GenomicFeatures)
library(GenomicAlignments)
library(DESeq2) # differential analysis
library(ChIPseeker) # annotation analysis
library(ggplot2)

# if using TxDb for annotation
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txDb <- TxDb.Hsapiens.UCSC.hg38.knownGene

# if using multi-cores
library(BiocParallel)
BiocParallel::register(BPPARAM=MulticoreParam(worker=4))

# package developted by CJW
# devtools::install_github('chaochaowong/peakable')
library(peakable)
```

Define directories:
```{r define-paths-and-variables}
# define directories
dir_header <- '///Volumes' # via MacBook
assoc_path <- '/data/hps/assoc/private/sarthy_lab'
result_dir <- file.path(assoc_path, 'NGS', 
                        'CUTnRUN_Results',
                        'GBe14_WSU-AML_CMS_FHD286_BRG1_Mods_CnT_20241028')

qc_dir <- file.path(result_dir, 'peak_qc')

# create R_proj folder
setwd(result_dir)
system('mkdir -p R_proj')
Rproj_dir  <- file.path(result_dir, 'R_proj')

# create data and figure folders
setwd(Rproj_dir)
system('mkdir -p data')
system('mkdir -p figures')
system('mkdir -p macs2_consensus_peaks')

data_dir           <- file.path(Rproj_dir, 'data')
consensus_peak_dir <- file.path(Rproj_dir, 'macs2_consensus_peaks')
fig_dir    <- file.path(Rproj_dir, 'figures')

# replicate group_by
rep_group_by  <- c('cell_line', 'antibody') # or treatment ....
```

__`sample_df`__ data frame: Import a sample sheet or construct a `data.frame` of sample information. The code chuck below imports an existing sample sheet (`.csv`) that includes, but not limited to `sample_id`, `cell_line`, `antibody` columns. Note that `sample_id` is mandated for `sample_df`.

```{r nf-sample-sheet-to-rds}
# import nexflow sample sheet: the sample sheet is expected to include but not limited
# to `sample_id`, `cell_line`, `antibody` columns.
# 
# - tidy up and save as 'sample_df.rds' and save to the data directory

sample_df <- read_csv(file.path(result_dir, 'data', 
                            'nf-sample-sheet.csv')) 

# tidy up and customize 'sample_df': exclude IgG
sample_df <- sample_df  %>%
  dplyr::filter(antibody != 'IgG') %>%
  dplyr::select(-single_end, -target_or_control, -read1, -read2)                            
saveRDS(sample_df, file.path(Rproj_dir, 'data', 'sample_df.rds'))
```

# MACS2

The code chunk below uses `peakable:::peakle_flow()` to import narrow/broad peak ranges (`grl`) and bam files from bed/bam files paths and extract sample metadata from the user provided _Nextflow_ sample sheet.

- `bam_pattern` was default fo '\\.markedDup.filter.sort.bam\$'     
- `bed_pattern` and `bam_pattern` must include characters that are subject to be removed for the files in order to restore the `sample_id` defined in  `sample_df$sample_id`
- The path to `bam` files is set to `result_dir/samtools_sort` (I might in the future allow user to define the `bam_dir`)


```{r get-macs2-peaks, eval=FALSE}
# use peakable:::peakle_flow() to import narrow/broad peak ranges from a NextFlow 
# result folder

bed_dir <- file.path(result_dir, 'peak_calls', 'macs2_callpeak')
bed_narrow_pattern <- '\\_peaks.narrowPeak$'
bed_broad_pattern  <- '\\_peaks.broadPeak$'
bam_dir <- file.path(result_dir, 'samtools_sort')
bam_pattern <- '\\.markedDup.filter.sort.bam$'

df_narrow <- sample_df %>% 
  dplyr::filter(!str_detect(antibody, 'K9me3|K27me3'))

df_broad <- sample_df %>% 
  dplyr::filter(str_detect(antibody, 'K9me3|K27me3'))

# make sure there is no IgG in antibody

# narrow: df_narrow, bed_narrow_pattern 
macs2_narrow <- # return: df and grl 
  peakable:::peakle_flow(sample_df = df_narrow, 
                         result_dir = result_dir,
                         peak_caller = 'MACS2-narrow',
                         peak_bed_dir = bed_dir, 
                         peak_bed_pattern = bed_narrow_pattern,
                         bam_dir = bam_dir,
                         bam_pattern = bam_pattern) 

macs2_broad <- # return: df and grl 
  peakable:::peakle_flow(sample_df = df_broad, 
                         result_dir = result_dir,
                         peak_caller = 'MACS2-broad',
                         peak_bed_dir = bed_dir, 
                         peak_bed_pattern = bed_broad_pattern,
                         bam_dir = bam_dir,
                         bam_pattern = bam_pattern) 

# combine narrow and broad peaks
macs2 <- peakable:::.macs2_bind(macs2_narrow, macs2_broad)

saveRDS(macs2, file=file.path(data_dir, 'macs2.rds'))
```

```{r macs-replicate-consensus, eval=FALSE}
# replicates consensus: group by the parameter "rep_group_by"
macs2_rep_consensus <- # use peakable:::consensus_by()
  peakable:::consensus_by(sample_df = macs2$df,
                          peaks_grl = macs2$grl,
                          consensus_group_by = rep_group_by,
                          peak_caller = 'macs2')

saveRDS(macs2_rep_consensus, 
        file=file.path(data_dir, 'macs2_rep_consensus.rds'))
```

```{r macs-consensus-summit-bed, message=FALSE, eval=FALSE}
# export macs2 consensus summit and narrowpeaks in bed and narrowPeak format, respectively, 
# to the R_proj/macs2_consensus_peaks sub-folder

# narrow peaks only

is_narrow <- function(peaks) {
  any(names(mcols(peaks)) == 'peak')
}

map2(macs2_rep_consensus$grl, names(macs2_rep_consensus$grl), 
     function(peaks, name) {
       # check if narrow peaks
         if (is_narrow(peaks)) {
           summit <- peakable::extract_summit_macs2(peaks)
           plyranges::write_bed(summit, 
                                file = file.path(consensus_peak_dir, 
                                                 paste0(name,
                                                        '_consensus_summit.bed')))
           plyranges::write_narrowpeaks(peaks, 
                                        file = file.path(consensus_peak_dir,
                                                         paste0(name,
                                                                '_cosensus_narrowPeak.bed')))
         }
       
       if (!is_narrow(peaks)) {
           # export consensus
           peakable::write_broadpeaks(peaks, 
                                      file = file.path(consensus_peak_dir,
                                                       paste0(name, '.broadPeak.bed')))
       }
})

```

```{r macs2-functional-annotation, message=FALSE, eval=FALSE}
# make sure the GRanges is not empty before using # ChIPseeker::anotatePeak()
keep <- macs2_rep_consensus$df %>%
  dplyr::filter(number_of_peaks > 0) %>% pull(sample_id)

macs2_annolst <- bplapply(macs2_rep_consensus$grl[keep],
                          ChIPseeker::annotatePeak,
                          tssRegion=c(-3000, 3000),
                          TxDb=txDb, annoDb="org.Hs.eg.db")
saveRDS(macs2_annolst, file=file.path(data_dir, 'macs2_annolst.rds'))
```
## Sample information
```{r load-peak-rds}
# load peaks 
macs2 <- readRDS(file.path(data_dir, "macs2.rds"))
macs2_rep_consensus <- readRDS(file.path(data_dir, 
                                         'macs2_rep_consensus.rds'))
```
Display a table of MACS2 sample, alignment, and peak information.

```{r display-macs2$df}
macs2$df %>%
  dplyr::select(sample_id, antibody, cell_line, 
                read_paired, dup_rate, peak_caller, number_of_peaks) %>%
  kable(caption='sample, peak, and samtools stats information') %>%
  kableExtra::kable_styling('striped')
```

# Peak numbers
Use `ggplot` to visualized peak numbers with the sample group, such as cell_line or antiboy.

```{r display-narrow-peak-number, fig.height=6, fig.cap='MACS2 narrow and broad peak numbers. X-axis (numbers) scalled by log10. Threshold: q-value = 0.01.'}
# viz peak number by bar plot
macs2$df %>%
  dplyr::arrange(antibody, cell_line) %>%
  dplyr::mutate(sample_id = factor(sample_id, levels=sample_id)) %>%
  ggplot(aes(y=sample_id, x=number_of_peaks, 
             color=antibody, fill=antibody)) +
    geom_bar(width=0.8, alpha = 0.8, stat="identity") +
    geom_text(aes(label=number_of_peaks), 
              vjust=0.5, hjust=1, color="white", size=3) +
    theme_minimal() +
    theme(panel.grid.major = element_blank(),
          legend.position = 'top')  +
  labs(title='MACS2 narrow and broad peak numbers') +
  scale_x_continuous(trans='log10') +
  theme(legend.position = 'none')
```

## PCA on peak hits

- Construct a n-by-m hit matrix of peaks against a collection of peakets, with entry 1 if a sample's peak hit the consolidated peaks and 0 otherwise. $n$ is the number of consolidated peaks and $m$ is the number of the samples
- Perform PCA on the hit matrix to obtain an overview of the peaks overlap/non-overlap variation across groups, histone marks, or treatments

```{r macs2-hit-mat}
# tokenize the hits on the corpus features and construct the hit matrix
macs2_hit_mat <- 
  peakable::consolidated_peak_hits(macs2$grl)
```

```{r macs2-pca-peak-hits}
# viz consolidated peak hits by PCA
# remove BRD9
macs2_hit_mat <- macs2_hit_mat[, colSums(macs2_hit_mat) > 0]

tmp <- macs2$df %>%
  dplyr::filter(sample_id %in% names(macs2_hit_mat)) %>%
  dplyr::mutate(antibody = factor(antibody))

macs2_hit_pca <- 
  peakable:::.getPCA(macs2_hit_mat, 
                      sample_info=tmp, n_pcs=2)
n_levels <- length(levels(tmp$antibody))

ggplot(macs2_hit_pca, aes(x=PC1, y=PC2, 
                          color=antibody, shape=antibody)) +
  geom_point() +
  theme_light() +
  facet_wrap(~cell_line, nrow=2) +
  scale_shape_manual(values=1:n_levels) +
  labs(title='PCA: MACS2 peak-hit')
```

# Replicate similarity

The code chuck below assumes the samples only have duplicates. You will have to modify the code if you have triplicates.

```{r cos-sim-by, fig.cap='Replicates cos similarity'}
sim_group_by = c('cell_line', 'antibody')
sim_df <- peakable::cos_similarity_by(sample_df = macs2$df,
                                      peaks_grl = macs2$grl,
                                      sim_group_by = sim_group_by)
```

```{r viz-macs-cos-sim}
sim_df %>%
  ggplot(aes(x=cos_sim, y=antibody)) +
    geom_point() +
    geom_segment(aes(x=0, y=antibody, xend=cos_sim, 
                     yend=antibody), color='grey50') +
    theme_light() +
    facet_wrap(~cell_line, nrow=2) +
    labs(title='replicates cos similarity')
```

# Consensus peaks

```{r macs2-consensus-peak-number}
# display macs2_rep_consensus
macs2_rep_consensus$df %>%
  dplyr::arrange(antibody, cell_line) %>%
  dplyr::mutate(sample_id = factor(sample_id, levels=sample_id)) %>%
ggplot(aes(x=number_of_peaks, y=sample_id,
       color=antibody, fill=antibody)) +
  geom_bar(width=0.8, alpha = 0.8, stat="identity") +
  geom_text(aes(label=number_of_peaks), 
            vjust=0.5, hjust=1, color="white", size=3) +
  theme_minimal() +
  theme(panel.grid.major = element_blank()) +
  labs(title='MACS2 replicates consensus peak numbers') +
  scale_x_continuous(trans='log2') +
  theme(legend.position = 'none')
```

# Functional annotation and feature distribution
- Use `ChIPseeker` to annotate the consensus peaks
- Plot feature distribution

```{r viz-macs2-functional-annotation, message=FALSE, fig.cap='Feature distribution of MACS2 consensus peaks for TFs', fig.width=6}
# make sure the GRanges is not empty before using # ChIPseeker::anotatePeak()
macs2_annolst <- readRDS(file.path(data_dir, 'macs2_annolst.rds'))
plotAnnoBar(macs2_annolst)
```


# SEACR 

```{r use-peak-flow-get-seacr-peaks}
# get import SEACR bed files 
bed_dir <- file.path(result_dir, 'peak_calls', 'seacr_callpeak')
bed_pattern <- '\\_norm.stringent.bed$'
#bed_pattern <- '\\_threshold0.01_norm.stringent.bed$'
bam_pattern <- '\\.markedDup.filter.sort.bam$'

#' use the peakle_flow to return seacr$df (info) and 
#' seacr$grl (peak ranges)
seacr_stringent <- 
  peakable:::peakle_flow(sample_df, # must be from nf_sample_sheet
                         result_dir = result_dir,
                         peak_caller = 'SEACR-stringent',
                         peak_bed_dir = bed_dir, 
                         peak_bed_pattern = bed_pattern,
                         bam_pattern = bam_pattern) 
# pick one per sample_id
saveRDS(seacr_stringent, file.path(data_dir, 'seacr_stringent.rds'))
```
