---
title: "[project_name] QC and downstream analysis"
author: "John Doe"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cosmo
    number_sections: true
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
    highlight: tango
    fig_caption: true
---


<script>
$(document).ready(function(){
    $('[data-toggle="popover"]').popover();
});
</script>

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = TRUE, 
                      echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE,
                      fig.align = 'center',
                      knitr.svg.object = TRUE,
                      dev=c('png', 'svglite'))
```

```{r load-libraries}
# load library
library(tidyverse)
library(plyranges)
library(GenomicRanges)
library(GenomicFeatures)
library(GenomicAlignments)
library(DESeq2) # differential analysis
library(ChIPseeker) # annotation analysis
library(ggplot2)

# if using TxDb for annotation
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txDb <- TxDb.Hsapiens.UCSC.hg38.knownGene

# if using multi-cores
library(BiocParallel)
#BiocParallel::register(BPPARAM=MulticoreParam(worker=4))

# package developted by CJW
# devtools::install_github('chaochaowong/peakalbe')
library(peakable)
```

```{r define-paths-and-variables}
# define directories
dir_header <- '///Volumes' # via MacBook
result_dir <- 
  file.path(dir_header, 'Active',
            'sarthy_j/Sarthy/Sarthy_Lab/CUTRUN_Records',
            'GBe4_M07e_palbo-50nM_FHD286-100nM_combo_CnT_04052024') 

Rproj_dir <- file.path(result_dir, 'R_proj')
data_dir <- file.path(Rproj_dir, 'data')
fig_dir <- file.path(Rproj_dir, 'figures')
```

```{r nf-sample-sheet-to-rds}
# - import nexflow sample sheet
# - tidy up and save as 'sample_df.rds' in the data directory
# - expect to have 'antibody' and 'cell_line' columns

# sample sheet should have 'antibody' and 'cell_line' column
sample_df <- read_csv(file.path(result_dir, 'data', 
                            'nf-sample-sheet.csv')) 

# tidy up and customize  'sample_df'
sample_df <- sample_df  %>%
  tidyr::drop_na(antibody) %>%
  dplyr::filter(!antibody == 'IgG') %>%
  dplyr::select(-single_end, -target_or_control, -read1, -read2)                            
saveRDS(sample_df, file.path(Rproj_dir, 'data', 'sample_df.rds'))
```

```{r load-and-tidy-sample-df}
#' sample_df is originated from the nextflow sample sheet
sample_df <- 
  readRDS(file.path(Rproj_dir, 'data', 'sample_df.rds')) 
```


# MACS2

The code chunk below uses `peakable:::peakle_flow()` to import peak ranges (`grl`) and information (`df`) of bed/bam files paths and sample metadata (extracted from the _Nextflow_ sample sheet). Note that

- `bam_pattern` was default fo '\\.markedDup.filter.sort.bam\$'     
- `bed_pattern` and `bam_pattern` must include characters that can be removed for the files in order to restore the `sample_id` defined in nextflow sample sheet (`sample_df\$sample_id`)      
- The path to `bam` files is set to `result_dir/samtools_sort` (I might in the future allow user to define the `bam_dir`)

```{r get-macs2-peaks}
# use peakle_flow
bed_dir <- file.path(result_dir, 'peak_calls', 'macs2_callpeak')
bed_pattern <- '\\_peaks.narrowPeak$'
#bam_dir <- file.path(result_dir, 'samtools_sort')
bam_pattern <- '\\.markedDup.filter.sort.bam$'

# exclude heterochromitan: H3K9me3 or H3K27me3
df_narrow <- sample_df %>% 
  dplyr::filter(!antibody %in% c('H3K9me3', 'H3K27me3'))

macs2_narrow <- # return: df and grl 
  peakable:::peakle_flow(sample_df = df_narrow, 
                         result_dir = result_dir,
                         peak_caller = 'MACS2-narrow',
                         peak_bed_dir = bed_dir, 
                         peak_bed_pattern = bed_pattern,
                         bam_pattern = bam_pattern) 

# broad peaks: K9me3 and(or) K27me3
df_broad <- sample_df %>% 
  dplyr::filter(antibody %in% c('H3K9me3', 'H3K27me3'))

macs2_broad <- # return: df and grl 
  peakable:::peakle_flow(sample_df = df_broad, 
                         result_dir = result_dir,
                         peak_caller = 'MACS2-broad',
                         peak_bed_dir = bed_dir, 
                         peak_bed_pattern = bed_pattern,
                         bam_pattern = bam_pattern) 

# combine narrow and broad peaks
macs2 <- peakable:::.macs2_bind(macs2_narrow, macs2_broad)
saveRDS(macs2, file=file.path(data_dir, 'macs2.rds'))
```

```{r macs-replicate-consensus}
# replicates consensus
macs2_rep_consensus <- # use peakable:::consensus_by()
  consensus_by(sample_df = macs2$df,
               peaks_grl = macs2$grl,
               consensus_group_by = c('treatment', 'Ab'),
               peak_caller = 'macs2')
saveRDS(macs2_rep_conensus, 
        file=file.path(data_dir, 'macs2_rep_consensus.rds'))
```

```{r load-peak-rds}
# load peaks 
macs2 <- readRDS(file.path(data_dir, "macs2.rds"))
macs2_consensus <- readRDS(file.path(data_dir, 
                                     'macs2_consensus.rds'))
```

```{r macs2-hit-mat}
# tokenize the hits on the corpus features and construct the hit matrix
macs2_hit_mat <- 
  peakable::consolidated_peak_hits(macs2$grl)
```

## Peak numbers
```{r display-narrow-peak-number, fig.height=6, fig.cap='MACS2 narrow and broad peak numbers. X-axis (numbers) scalled by log2. Threshold: q-value = 0.01.'}
# viz peak number by bar plot
macs2$df %>%
  dplyr::arrange(antibody, cell_line) %>%
  dplyr::mutate(sample_id = factor(sample_id, levels=sample_id)) %>%
  ggplot(aes(y=sample_id, x=number_of_peaks, 
             color=antiboy, fill=antibody)) +
    geom_bar(width=0.8, alpha = 0.8, stat="identity") +
    geom_text(aes(label=number_of_peaks), 
              vjust=0.5, hjust=1, color="white", size=3) +
    theme_minimal() +
    theme(#panel.grid.minor = element_blank(),
          panel.grid.major = element_blank(),
          legend.position = 'top')  +
  labs(title='MACS2 narrow peak numbers') +
  scale_x_continuous(trans='log2') +
  theme(legend.position = 'none')
```

## PCA on peak hits

- Construct a n-by-m hit matrix of peaks against a collection of peakets, with entry 1 if a sample's peak hit the consolidated peaks and 0 otherwise. $n$ is the number of consolidated peaks and $m$ is the number of the samples
- Perform PCA on the hit matrix to obtain an overview of the peaks overlap/non-overlap variation across groups, histone marks, or treatments

```{r macs2-pca-peak-hits}
# viz consolidated peak hits by PCA
# remove BRD9
macs2_hit_mat <- macs2_hit_mat[, colSums(macs2_hit_mat) > 0]

tmp <- macs2$df %>%
  dplyr::filter(sample_id %in% names(macs2_hit_mat)) %>%
  dplyr::mutate(antibody = factor(antibody))

macs2_hit_pca <- 
  peakable:::.getPCA(macs2_hit_mat, 
                      sample_info=tmp, n_pcs=2)
n_levels <- length(levels(tmp$antibody))

ggplot(macs2_hit_pca, aes(x=PC1, y=PC2, 
                          color=anbibody, shape=antibody)) +
  geom_point() +
  theme_light() +
  facet_wrap(~cell_line, nrow=2, scales='free') +
  scale_shape_manual(values=1:n_levels) +
  labs(title='PCA: MACS2 peak-hit')
```

## Replicate similarity

The code chuck below assumes the samples only have duplicates. You will have to modify the code if you have triplicates.

```{r macs-cos-sim}
#' (1) build hit mat between duplicates
#' (2) cos similarity
cos_sim <- macs2$df %>%
  group_split(antibody, cell_line) %>%
  map_dfr(function(x) {
    if (any(x$number_of_peaks == 0))
      sim <- 0
    
    if (all(x$number_of_peaks > 0)) {
      hit_map <- 
        peakable::consolidated_peak_hits(macs2$grl[x$sample_id])
      sim <- peakable:::cos_similarity(x=hit_map[, 1, drop=TRUE],
                                        y=hit_map[, 2, drop=TRUE])
    }
    
    data.frame(antibody = x$antibody[1],
               cell_line = x$cell_line[1],
               name = paste(x$sample_id, collapse = '-vs-'),
               cos_sim = sim)
  })
```

```{r viz-macs-cos-sim}
cos_sim %>%
  ggplot(aes(x=cos_sim, y=antibody)) +
    geom_point() +
    geom_segment(aes(x=0, y=antibody, xend=cos_sim, 
                     yend=antibody), color='grey50') +
    theme_light() +
    facet_wrap(~cell_line, nrow=2) +
    labs(title='cos similarity')
```

## Consensus peaks
Display consensus peak numbers

```{r macs2-consensus-peak-number}
df <- data.frame(sample_id = names(macs2_consensus),
                 number_of_consensus_peaks= 
                   sapply(macs2_consensus, length)) %>%
  dplyr::mutate(Ab = 
                  str_split(sample_id, "-", simplify=TRUE)[, 1],
                cell_line = 
                  str_split_fixed(sample_id, "-", n=2)[, 2]) %>%
  dplyr::arrange(Ab, cell_line) %>%
  dplyr::mutate(sample_id = factor(sample_id, levels=sample_id))

ggplot(df, aes(x=number_of_consensus_peaks, y=sample_id,
               color=Ab, fill=Ab)) +
  geom_bar(width=0.8, alpha = 0.8, stat="identity") +
  geom_text(aes(label=number_of_consensus_peaks), 
            vjust=0.5, hjust=1, color="white", size=3) +
  theme_minimal() +
  theme(panel.grid.major = element_blank()) +
  labs(title='MACS2 replicates consensus peak numbers') +
  scale_x_continuous(trans='log2') +
  theme(legend.position = 'none')
```

## Functional annotation and feature distribution
- Use `ChIPseeker` to annotate the consensus peaks
- Plot feature distribution

```{r macs2-functional-annotation, message=FALSE, fig.cap='Feature distribution of MACS2 consensus peaks for TFs'}
# how many known genes
# how many unique genes' TSS it binds to?
macs2_annolst_M07e <- bplapply(macs2_consensus,
                     ChIPseeker::annotatePeak,
                     tssRegion=c(-3000, 3000),
                     TxDb=txDb, annoDb="org.Hs.eg.db")

plotAnnoBar(macs2_annolst_M07e)
```


# SEACR

Code chunks in this section would be mostly similar to the previous one.
```{r use-peak-flow-get-seacr-peaks}
# get import SEACR bed files 
bed_dir <- file.path(result_dir, 'peaks_calls', 'seacr_callpeak')
bed_pattern <- '\\_threshold0.01_non.stringent.bed$'
# bam_pattern <- '\\.markedDup.filter.sort.bam$'
bam_pattern <- '\\.rmDup.filter.sort.bam$'

#' use the peakle_flow to return seacr$df (info) and 
#' seacr$grl (peak ranges)
seacr <- 
  peakable:::peakle_flow(sample_df, # must be from nf_sample_sheet
                          result_dir = result_dir,
                          peak_caller = 'SEACR-thres1p',
                          peak_bed_dir = bed_dir, 
                          peak_bed_pattern = bed_pattern,
                          bam_pattern = bam_pattern) 
# pick one per sample_id
saveRDS(seacr, file.path(data_dir, 'seacr_thres1p.rds'))
```
